File: apiendpoint/function_compute_handler.py
"""
Function Compute Handler for FastAPI
Handles HTTP and event-based invocations
"""

import json
import logging
from typing import Any, Dict
from app.main import app

logger = logging.getLogger('fc_handler')

async def http_handler(request, context) -> Any:
    """
    HTTP trigger handler
    Receives HTTP requests from API Gateway or direct invocation
    
    Args:
        request: HTTP request object
        context: Function Compute context
        
    Returns:
        HTTP response
    """
    try:
        # FastAPI/Starlette handles the request
        return await app(request, context)
    except Exception as e:
        logger.error(f"HTTP handler error: {e}", exc_info=True)
        return {
            'statusCode': 500,
            'body': json.dumps({
                'error': 'Internal server error',
                'message': str(e)
            })
        }

async def event_handler(event: Dict[str, Any], context: Any) -> Dict[str, Any]:
    """
    Event trigger handler
    Handles async events from MQ, Timer, etc.
    
    Args:
        event: Event data
        context: Function Compute context
        
    Returns:
        Event processing result
    """
    event_type = event.get('type')
    
    logger.info(f"Processing event: {event_type}")
    
    try:
        if event_type == 'production_order_created':
            # Handle production order creation
            return {
                'success': True,
                'message': 'Production order event processed',
                'event_type': event_type
            }
        
        elif event_type == 'qc_inspection_completed':
            # Handle QC inspection completion
            return {
                'success': True,
                'message': 'QC inspection event processed',
                'event_type': event_type
            }
        
        elif event_type == 'delivery_created':
            # Handle delivery creation
            return {
                'success': True,
                'message': 'Delivery event processed',
                'event_type': event_type
            }
        
        else:
            logger.warning(f"Unknown event type: {event_type}")
            return {
                'success': False,
                'message': f"Unknown event type: {event_type}"
            }
    
    except Exception as e:
        logger.error(f"Event handler error: {e}", exc_info=True)
        return {
            'success': False,
            'error': str(e)
        }

# Export handlers for Function Compute
__all__ = ['http_handler', 'event_handler', 'app']

File: apiendpoint/node/function-compute-handler.js
/**
 * Function Compute Handler for Alibaba Cloud
 * Handles function invocations from Alibaba FC
 */

const app = require('./server-express');

/**
 * HTTP function handler for Alibaba Function Compute
 * Automatically invoked for HTTP triggers
 * 
 * @param {Object} req - HTTP request object
 * @param {Object} res - HTTP response object
 */
async function handler(req, res) {
    try {
        // Log incoming request
        console.log(`[FC Handler] ${req.method} ${req.path}`);
        
        // Pass to Express app
        return app(req, res);
    } catch (error) {
        console.error('[FC Handler] Error:', error);
        res.status(500).json({
            error: 'Internal Server Error',
            message: error.message
        });
    }
}

/**
 * Event-based handler for Alibaba Function Compute
 * Used for asynchronous processing
 * 
 * @param {Object} event - Event data
 * @param {Object} context - Function context
 * @returns {Promise}
 */
async function eventHandler(event, context) {
    try {
        console.log('[FC Event Handler] Processing event:', event);
        
        // Process event based on type
        const { eventType, data } = event;
        
        switch (eventType) {
            case 'production_order_created':
                console.log('[FC Event Handler] Processing production order creation');
                // Handle production order creation
                break;
                
            case 'qc_inspection_completed':
                console.log('[FC Event Handler] Processing QC inspection');
                // Handle QC completion
                break;
                
            case 'delivery_created':
                console.log('[FC Event Handler] Processing delivery creation');
                // Handle delivery creation
                break;
                
            default:
                console.log('[FC Event Handler] Unknown event type:', eventType);
        }
        
        return {
            statusCode: 200,
            body: JSON.stringify({
                message: 'Event processed successfully',
                eventType: eventType
            })
        };
    } catch (error) {
        console.error('[FC Event Handler] Error:', error);
        return {
            statusCode: 500,
            body: JSON.stringify({
                error: 'Event processing failed',
                message: error.message
            })
        };
    }
}

module.exports = {
    handler,      // HTTP function handler
    eventHandler, // Event-based handler
    app          // Express app for testing
};
